### DISCUSSION TITLE
Expand with index (python enumerate style). What is the best way to do it?

### DISCUSSION THREAD

**Discussion by emcastro** [dflib/dflib#420] [Q&A] [answered]
Let's take the classical example with phone numbers and expansion on phones:
```java
DataFrame df = DataFrame.foldByRow("name", "phones").of(
        "Cosin", List.of("111-555-5555","111-666-6666","111-777-7777"),
        "O'Hara", List.of("222-555-5555"));

df.rows()
  .selectExpand("phones");
```
```
name   phones      
------ ------------
Cosin  111-555-5555
Cosin  111-666-6666
Cosin  111-777-7777
O'Hara 222-555-5555
4 rows x 2 columns
```

First question: It is unclear to me how to easily rename `phones` to `phone` on the fly. The following seems a little cumbersome:
```java
df.rows()
  .selectExpand("phones")
  .cols().selectAs(Map.of("phones", "phone"))
```

Then the main point of this discussion. What I want to get is:
```
name   phone        phone_id
------ ------------ --------
Cosin  111-555-5555        1
Cosin  111-666-6666        2
Cosin  111-777-7777        3
O'Hara 222-555-5555        1
4 rows x 3 columns
```
It can be easily achieved with group-by, but it is unfortunate that an additional iteration on the column be required to calculate `phone_id` as the information seems to be already available in the expand process.
```java
df.rows()
  .selectExpand("phones")
  .group("name")
  .cols("name","phone","phone_id").select($col("name"), $col("phones"), rowNum())
```

Another option would be to zip phone numbers with their id, but it is really cumbersome:
```java
public List<List> enumerate(List list) {
    var res = new ArrayList<List>(list.size());
    int i = 1;
    for (var data : list) {
        res.add(List.of(i, data));
        i++;
    }
    return res;
}

// enumerate(List.of("a","b","c")) => [[0, a], [1, b], [2, c]]

df.cols("phones").merge(Exp.$col("phones",List.class).mapVal(p->enumerate(p)))
```
```
name   phones                                            
------ --------------------------------------------------
Cosin  [[1, 111-555-5555], [2, ..666], [3, 111-777-7777]]
O'Hara [[1, 222-555-5555]]                               
2 rows x 2 columns
``` 
```java
df.cols("phones").merge(Exp.$col("phones",List.class).mapVal(p->enumerate(p)))
    .rows().expand("phones")
    .cols("phone_id","phone").expand($col("phones"))
    .cols("name", "phone", "phone_id").select()
```
```
name   phone        phone_id
------ ------------ --------
Cosin  111-555-5555        1
Cosin  111-666-6666        2
Cosin  111-777-7777        3
O'Hara 222-555-5555        1
4 rows x 3 columns
```

For my first question a nice syntax would be:
```java
df.rows().selectExpandAs("phones","phone")
```
or
```java
df.cols("phone").rows().selectExpand("phones")
```
Note: the _select_ applies to the rows, not the columns.

For the main question:
```java
df.rows().selectEnumerateAs("phones", "phone", "phone_id");
```
or
```java
df.cols("phone", "phone_id").rows().selectEnumerate("phones");
```

**Comment by andrus** [comment, marked as answer]
Good points. Collapsing multiple operations/transformations into a single row/column set run is something I totally support. But it is also a question of clear semantics (how to make the API both unambiguous and minimally verbose for simpler cases). So I suspect the direction we'll need to take is to separate the expansion step (for rows, but also columns) into an intermediate method (that returns a row /column set instead of the final `DataFrame` result) . Smth like this:

```java
df.rows().expand("phone").cols("name", "phone", "id").select($col("name"), $col("phones"), rowNum());
```

**Reply by emcastro** [reply]
Your proposition solves both problems at the same time. 👍

**Comment by andrus** [comment]
Cool. I created issues for those tasks:

* https://github.com/dflib/dflib/issues/421
* https://github.com/dflib/dflib/issues/422

IIRC, initially all the varieties of the `extend(..)` method were made terminal as they change the geometry of row/column sets, something we were not yet ready to handle. I think we can do it now.
